### ISSUE TITLE
Gradle should not silently allow two subprojects with the same name (creating the same artifact)

### ISSUE THREAD

**Issue by C-Otto** [gradle/gradle#16986]
This is a continuation of #16885.
Imagine a developer creates two subprojects (`:bbb` and `:aaa:bbb` in my example), which share the same group and name, but are intended to be different. Currently, gradle treats them as "identical" in some sense, and only uses one out of the two as part of dependency resolution.

In a simple project where `:aaa` includes both `:bbb` and `:aaa:bbb`, Gradle decides on one of the two implementations, without a warning/error that this is ambigous and possibly unintentional. As a result, a developer might not even know that one of the subprojects is completely ignored due to dependency resolution. In case of cyclic dependencies that only result out of the identification of the two subprojects with the same name, the cyclic dependency error is hard to understand (see #16885).

### Expected Behavior
I'd like Gradle to fail the build or at least emit a warning such that developers can work around the issue.

### Current Behavior
Gradle silently picks one of the two implementations, causing issues like #16885 and other confusion.

### Context
Reproducer: https://github.com/C-Otto/gradle-nested-same-name/tree/without-cycle

**Comment by C-Otto** [comment]
@Vampire added (in slack):

> I think if you try to publish there might already be such a warning, but would be that this was just within one project where multiple publications publish the same artifact.

**Comment by C-Otto** [comment]
Removing the group (which is set for all (nested) subprojects) helps.

```
$ git diff
diff --git a/build.gradle b/build.gradle
index a211884..efd3a45 100644
--- a/build.gradle
+++ b/build.gradle
@@ -1,4 +1,3 @@
 subprojects {
-    group 'de.cotto'
     version '1.0-SNAPSHOT'
-}
\ No newline at end of file
+}
```

**Comment by Vampire** [comment]
This is related to #847

**Comment by ljacomet** [comment]
While fixing the global issue of project identity is tricky, Gradle should be capable of failing (or at least warning) when dependency resolution conflict resolves two projects with the same GAV coordinates but different project paths.

**Comment by rvplauborg** [comment]
I agree. I just spent a lot of the day debugging as I ran into this it seems...

**Comment by ZacSweers** [comment]
It's a real issue

**Comment by illarionov** [comment]
@mlopatkin mentioned in https://github.com/gradle/gradle/issues/17784#issuecomment-883991005 that there are plans to add a warning about duplicate project names. Any progress on this?
This issue, along with #847, is quite old and among the most upvoted. This behavior of Gradle is confusing for most developers and commonly encountered. At the same time, it is not even mentioned in the documentation.

**Comment by SgtSilvio** [comment]
>> While fixing the global issue of project identity is tricky
>
>Do we need to fix this to avoid the cycle? They seem 2 different problems to me.
>The worse part of this to me is that, by changing the name of the project, you also change the path (!!)

I agree with this, project path and project name should be independently configurable. I wrote a longer explanation here: https://github.com/gradle/gradle/issues/847#issuecomment-2833435762

**Comment by aSemy** [comment]
I have proposed a fix in #39241.

**Comment by bcmedeiros** [comment]
I don't think we should warn of fail the build, we should pave the way to support this. The whole point of having the :a:b notation is it be able to add a :z:b later without impacting existing modules.

This is not mentioned in the documentation because it was never meant to work like this, it was not a designed feature, it was accidental. And now that it's "too hard" to fix, we are pretending it is intended behaviour and adding validations?

**Comment by Vampire** [comment]
@bcmedeiros I don't get what you mean.
Yes, projects with same name should be and are supported, as long as their coordinates (i.e. group + name) which is their identity in terms of artifacts is unique.
But if two projects produce the same group + name, then this is a build bug and should fail the build cleanly instead of picking one of the two for all dependencies.

**Comment by Vampire** [comment]
Well, as long as the projects are not published and there is no dependency to these projects, there is probably not a problem **yet** and it would not necessarily need to fail the build.

But latest when there is a dependency added to one of the projects or they are published, it should fail cleanly with a meaningful error instead of just picking one, or overwriting each others publication artifacts.

So it is probably easier and safer to right away prevent two projects with the same group + name in the same build.
